cmd/kube-apiserver-healthcheck/main.go

```diff
@@ -0,0 +1,164 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"flag"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"os"
+
+	"k8s.io/klog"
+)
+
+// healthCheckServer is the http server
+type healthCheckServer struct {
+	transport *http.Transport
+}
+
+// handler processes a single http request
+func (s *healthCheckServer) handler(w http.ResponseWriter, r *http.Request) {
+	if r.Method == "GET" {
+		switch r.URL.Path {
+		case "/.kube-apiserver-healthcheck/healthz":
+			// This is a check for our own health
+			w.WriteHeader(http.StatusOK)
+			w.Write([]byte("ok"))
+			return
+
+		case "/healthz", "/readyz":
+			// This is a health-check we will proxy
+			s.proxyRequest(w, r)
+			return
+		}
+	}
+
+	klog.Infof("unknown request: %s %s", r.Method, r.URL.Path)
+	http.Error(w, "not found", http.StatusNotFound)
+}
+
+// httpClient builds an isolated http.Client
+func (s *healthCheckServer) httpClient() *http.Client {
+	return &http.Client{Transport: s.transport}
+}
+
+// proxyRequest forwards a request, once it has been verified by handler
+func (s *healthCheckServer) proxyRequest(w http.ResponseWriter, r *http.Request) {
+	httpClient := s.httpClient()
+
+	u := &url.URL{
+		Scheme:   "https",
+		Host:     "127.0.0.1",
+		Path:     r.URL.Path,
+		RawQuery: r.URL.RawQuery,
```

Comment:
Do you really want to pass through an unvetted `RawQuery`? It seems risky to me.

Reply:
So it's not trivial.. https://github.com/kubernetes/kubernetes/blob/08e1fd3bb947faf465e8a67d5c7106dbd10840c0/cluster/gce/manifests/kube-apiserver.manifest#L39

I don't know what validation we could / should do here.  We could whitelist some queries I guess but I think that'll confuse people more.  I imagine two things are coming:

1) exposing this for use as a load balancer health-check (i.e. publicly)
2) using the richer readyz / livez queries as in the upstream manifest

Reply:
I think we need to parse the query and only pass through parameters we recognize as safe.

Reply:
Looking at apiserver code, I see at least `exclude` and `verbose`, but that `http.Request` gets passed around all over the place.

So it's a risk assessment as to how likely it is that apiserver would have or later add an important authorization check on a parameter to those paths.

Reply:
How about we parse and pass through `exclude`? I worry about information leakage from `verbose`.

Reply:
Great idea - I implemented that.  I also made it more testable and added a test!